### PR TITLE
 macOS: end-user postinstall script to parse promo code from installer .pkg filename

### LINKED ISSUE #12786: Script to read executable name (PKG) and output to profile directory
## Test plan

https://github.com/brave/browser-laptop/pull/12953#issue-293024981

----

**NOTE: https://github.com/brave/browser-laptop/issues/12784 must be completed first**

During the PKG file install for a Release channel build (not BETA/DEVELOPER/NIGHTLY), a custom script should run. This script needs to do the following:
- Read executable name (ex: `Brave-ABC.pkg`)
- Look for / parse out a promotion code (ex: `ABC`)
- Put this promotional code into a file `promoCode` in the profile directory  (ex: `~/Library/Application Support/brave`)

### PR DESCRIPTION
**Pull request by petemill**
Bash script parses promocode e.g. pem001 from filename e.g. /path/to/Brave-Setup-v0.22-pem001.pkg

Fix #12786 

**Depends on #12946 (use pkg installer instead of dmg for macOS release channel)** ✅ now merged

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

### Test plan
- Delete all Brave.app's everywhere (/Applications, browser-laptop/Brave*/, ...)
  - the .pkg installer will see these as existing (and moved installations), and upgrade them instead)
- Build package
- Build installer
- Rename installer with `-pem001` at end (before .pkg)
- Run installer

Observe:
- Brave still runs correctly after install
- `promoCode` file is created inside user-data dir with the contents: `pem001`

Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests

- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

### CODE REVIEW

File: build/pkg-scripts/postinstall [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
+# This postinstall script is run on macOS pkg installers
+
+# Extract a referral / promo code from the installer path of the format /dir/path/Setup-Brave-Anything-xxx001.pkg
+# where the promo code would be xxx001
+installerPath=$1
+installerPathPromoCodeRegex='.+([a-zA-Z0-9]{3}[0-9]{3})\.pkg$'
```
Comment (diracdeltas) [inline build/pkg-scripts/postinstall]: should this be `.+-([a-zA-Z0-9]{3}[0-9]{3})\.pkg$` to match the windows regex?

Comment (petemill) [reply build/pkg-scripts/postinstall]: got spot. `-` went missing during testing the different syntax

File: build/pkg-scripts/postinstall [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
+# This postinstall script is run on macOS pkg installers
+
+# Extract a referral / promo code from the installer path of the format /dir/path/Setup-Brave-Anything-xxx001.pkg
+# where the promo code would be xxx001
+installerPath=$1
+installerPathPromoCodeRegex='.+-([a-zA-Z0-9]{3}[0-9]{3})\.pkg$'
+userDataDir="$HOME/Library/Application Support/brave"
```
[the remark this answers is not in the document]
Comment (bbondy) [reply build/pkg-scripts/postinstall]: I don't think it's important, but just noting this won't work for different channel builds.

[the remark this answers is not in the document]
Comment (NejcZdovc) [reply build/pkg-scripts/postinstall]: currently on my mac `brave` is lowercase

[the remark this answers is not in the document]
Comment (petemill) [reply build/pkg-scripts/postinstall]: confirmed, is lower-case on v0.20 and up

**Comment by petemill** [comment]
0.22.x eafdb4f15
0.21.x 28ff27e3c
0.20.x a1726d613

**Pull request merged**